### PR TITLE
Use agent directly

### PR DESCRIPTION
**Pull request by cviecco**

### CODE REVIEW

File: cmd/keymaster/agent.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+
+	"github.com/Cloud-Foundations/Dominator/lib/log"
+	"github.com/cviecco/npipe"
```
Comment (rgooch) [inline cmd/keymaster/agent.go]: Please don't add a dependency on a personal repository.

File: cmd/keymaster/agent.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		return err
+	}
+
+	keyToAdd := agent.AddedKey{
+		PrivateKey:  privateKey,
+		Certificate: sshCert,
+		Comment:     comment,
+	}
+	// NOTE: Current Windows ssh (OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5)
+	// barfs when encountering a lifetime so we only add it for non-windows
```
Comment (rgooch) [inline cmd/keymaster/agent.go]: Doesn't this mean that `ssh-agent` will fill up with keys? What's the behaviour with the old `ssh-add` approach?

Comment (cviecco) [reply cmd/keymaster/agent.go]: It is similar to this, it deletes the entries that match the keyname. What will happen is that the key will not autoexpire from windows.

File: cmd/keymaster/agent.go (hunk 1 above)
Comment (cviecco) [reply cmd/keymaster/agent.go]: I can fork it into cloud-foundations. Would this work for you?

Comment (rgooch) [reply cmd/keymaster/agent.go]: That would be better. Is the upstream maintainer non-responsive?

File: cmd/keymaster/agent.go (hunk 2 above)
Comment (rgooch) [reply cmd/keymaster/agent.go]: OK, so just to be clear: with this new code, the agent will not collect an increasing number of stale keys?

File: cmd/keymaster/agent.go (hunk 1 above)
Comment (cviecco) [reply cmd/keymaster/agent.go]: Seems so, there is no activity for the las 3 years and pull requests waiting. For now a fork withing cloud-foundations seems better. The only change that I did was make a stub for all non-windows OS. Want to review that?  Of should I just land?

File: cmd/keymaster/agent.go (hunk 2 above)
Comment (cviecco) [reply cmd/keymaster/agent.go]: Correct. There will be at most 1 stale cert on windows.

File: cmd/keymaster/agent.go (hunk 1 above)
Comment (cviecco) [reply cmd/keymaster/agent.go]: Also, while we agree on this there will be another push with this item pending.

**Comment by rgooch** [comment]
Cut&pasting one of my original comments:
How about moving this code to the golib repository? At the very least it should be a separate package within the Keymaster repository.

**Comment by cviecco** [comment]
Eventually yes... but I want more at least one more running use case before I can start solidifying the api. Maybe an experimental x/golib library? I would be fine with that witht he code in its current state.

**Comment by rgooch** [comment]
You've only got three straightforward function entry points. Take a quick stock of the parameters and if you can't think of any others you might want, it's good enough for now. You can always add new functions with `2` suffixed.

If we create a "experimental" tree, I fear it will just collect stuff that just gets dumped and never cleaned up/moved. It will be work to go back and do that later. Also, it just makes it easier to write the minimum to get the job done without even thinking how it might be used by others. I find merely thinking about how a package might be used by others improves the quality of the code, even if many features are not implemented.

File: cmd/keymaster/agent.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import (
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+
+	"github.com/Cloud-Foundations/Dominator/lib/log"
```
Comment (rgooch) [inline cmd/keymaster/agent.go]: s@Dominator/lib/log@golib/pkg/log@

**Pull request merged**